client/consensus/grandpa/environment_test.go

```diff
@@ -0,0 +1,687 @@
+	enc, err := scale.Marshal(compRounds)
+	require.NoError(t, err)
+
+	var newCompletedRounds completedRounds[string, uint]
+	err = scale.Unmarshal(enc, &newCompletedRounds)
+	require.NoError(t, err)
+	require.Equal(t, compRounds, newCompletedRounds)
+}
+
+func TestCompletedRounds_Last(t *testing.T) {
+	authorities := AuthoritySet[string, uint]{
+		CurrentAuthorities:     []Authority{},
+		SetID:                  1,
+		PendingStandardChanges: NewChangeTree[string, uint](),
+		PendingForcedChanges:   []PendingChange[string, uint]{},
+		AuthoritySetChanges:    AuthoritySetChanges[uint]{},
+	}
+
+	dummyHashNumber := grandpa.HashNumber[string, uint]{
+		Hash:   "a",
+		Number: 1,
+	}
+
+	completedRound := completedRound[string, uint]{
+		Number: 1,
+		State: grandpa.RoundState[string, uint]{
+			PrevoteGHOST: &dummyHashNumber,
+			Finalized:    &dummyHashNumber,
+			Estimate:     &dummyHashNumber,
+			Completable:  true,
+		},
+		Base: dummyHashNumber,
+	}
+	completedRounds := NewCompletedRounds[string, uint](&completedRound, 1, authorities)
+	lastCompletedRound := completedRounds.last()
+	require.Equal(t, completedRound, lastCompletedRound)
+
+	completedRounds = NewCompletedRounds[string, uint](nil, 1, authorities)
+	require.Panics(t, func() { completedRounds.last() }, "last did not panic")
+}
+
+func TestCompletedRounds_Push(t *testing.T) {
+	authorities := AuthoritySet[string, uint]{
+		CurrentAuthorities:     []Authority{},
+		SetID:                  1,
+		PendingStandardChanges: NewChangeTree[string, uint](),
+		PendingForcedChanges:   []PendingChange[string, uint]{},
+		AuthoritySetChanges:    AuthoritySetChanges[uint]{},
+	}
+
+	dummyHashNumber := grandpa.HashNumber[string, uint]{
+		Hash:   "a",
+		Number: 1,
+	}
+
+	completedRound0 := completedRound[string, uint]{
+		Number: 1,
+		State: grandpa.RoundState[string, uint]{
+			PrevoteGHOST: &dummyHashNumber,
+			Finalized:    &dummyHashNumber,
+			Estimate:     &dummyHashNumber,
+			Completable:  true,
+		},
+		Base: dummyHashNumber,
+	}
+
+	completedRound1 := completedRound[string, uint]{
+		Number: 1,
+		State: grandpa.RoundState[string, uint]{
+			PrevoteGHOST: &dummyHashNumber,
+			Finalized:    &dummyHashNumber,
+			Estimate:     &dummyHashNumber,
+			Completable:  true,
+		},
+		Base: dummyHashNumber,
+	}
+
+	completedRound2 := completedRound[string, uint]{
+		Number: 1,
+		State: grandpa.RoundState[string, uint]{
+			PrevoteGHOST: &dummyHashNumber,
+			Finalized:    &dummyHashNumber,
+			Estimate:     &dummyHashNumber,
+			Completable:  true,
+		},
+		Base: dummyHashNumber,
+	}
+	completedRounds := NewCompletedRounds[string, uint](&completedRound1, 1, authorities)
+	completedRounds.push(completedRound0)
+
+	lastCompletedRound := completedRounds.last()
+	require.Equal(t, completedRound0, lastCompletedRound)
+
+	completedRounds.push(completedRound2)
+	lastCompletedRound = completedRounds.last()
+	require.Equal(t, completedRound0, lastCompletedRound)
+}
```

Comment:
This test doesn't actually test any of the sorting logic.  `completedRound0` is the same as `completedRound1` and `completedRound2`.  I think you should write the same test with differing rounds in rust as well, and make sure we're doing the same thing as the rust type.

Reply:
This test has been updated, will verify with respect to substrate

Reply:
After spending more time reading the substrate code, I realized that I needed to change how these were ordered. I have updated the logic as well as the test.